### ISSUE TITLE
Split TS declaration files are not compatible with nodenext module resolution

### ISSUE THREAD

**Issue by ssangervasi** [developit/microbundle#1019]
TLDR: Microbundle should output a single `.d.ts` per output module file, _or_ `nodenext` is too unstable to target at all.

### Background
This issue is a funky mixture of a few other issues:
1. The ongoing changes to TS's module resolution, where `nodenext` (aka `node16`) uses modules: https://github.com/microsoft/TypeScript/issues/50152
2. Microbundle's basic documentation of `nodenext` compat: https://github.com/developit/microbundle/issues/990
3. And the fact microbundle generates multiple `.d.ts` files for a single `.*js` output: https://github.com/developit/microbundle/issues/239

### Reproducing
I forked the code from #239 and updated it to highlight the problem with nodenext: https://github.com/ssangervasi/microbundle-typings. 
In this example, you can see:
1. :heavy_check_mark: The "library" package is built with microbundle
2. :heavy_check_mark: The library package has `exports.types` pointing to the module's types, `dist/index.d.ts`, as suggested by microbundle's readme.
3. :neutral_face: That declaration file has a relative dependency on another file produced by microbundle.
4. :heavy_check_mark: The dependent package is built with `tsc 4.9.4` with `nodenext`. _(This is the thing I want the users of my package to be able to do.)_
5. :heavy_check_mark: The dependent package can resolve the library package's root declaration.
6. :disappointed: TypeScript can't resolve the relative dependency produced by microbundle.

### Proposed solution
My instinct is that generating multiple `.d.ts` files for a single output module is going to cause problems in general. In this case, if the build process could merge them (the same way it merges the JS into a single file), the TS resolution problem would go away. Basically reopening https://github.com/developit/microbundle/issues/239.

On the other hand, TS's `moduleResolution` is madenning*, and it's possible this relative lookup will go away in later versions. Microbundle's readme could be updated to note this incompatibility for now. As far as I can tell, there isn't a way to get these split `.d.ts` files to work, except maybe [package.json redirects](https://github.com/andrewbranch/example-subpath-exports-ts-compat/tree/main/examples/node_modules/package-json-redirects) which I'm not willing to try or publish.

\* On par with how complex JS's dependency system is.

**Comment by rschristian** [comment]
TLDR: there's an easy fix:

```diff
// src/index.ts
-export * from "./foo";
+export * from "./foo.js";
```

Long answer: this is a TS issue, not a Microbundle issue. We use a (few times) wrapped `tsc` to handle TypeScript and this is the way the TypeScript team have chosen to go. Yes, most of us dislike this (using `.js` to refer to `.ts` or `.tsx` files), but that's what we're stuck with as they've drawn a line in the sane.

Merging `.d.ts` files is much more complicated and breaks from native `tsc` behavior, so I don't think it's worth pursuing (although it is a nice feature).

**Comment by ssangervasi** [comment]
Holy moly, I [read](https://github.com/microsoft/TypeScript/issues/50152) about that but still couldn't understand the fix applied to this situation.

> TypeScript will impose Node’s much stricter ESM resolution algorithm on those files, disabling index-file resolution and extensionless lookups—in fact, the extension the user has to write is .js, which will be nonsensical for the context, where the runtime module resolver (the bundler) only ever sees .ts files. 

Thanks for explaining!

**Comment by rschristian** [comment]
No problem, it's certainly a weird one. Using `foo.js` to refer to `foo.d.ts` file is quite the odd design choice to most users I think.
